crates/ark/src/lsp/completions/sources/composite.rs

```diff
@@ -17,116 +17,122 @@ mod workspace;
 use std::collections::HashSet;
 
 use anyhow::Result;
-use call::completions_from_call;
-use document::completions_from_document;
-use keyword::completions_from_keywords;
-use pipe::completions_from_pipe;
-use pipe::find_pipe_root;
-use search_path::completions_from_search_path;
-use snippets::completions_from_snippets;
+pub use pipe::find_pipe_root;
 use stdext::*;
-use subset::completions_from_subset;
 use tower_lsp::lsp_types::CompletionItem;
 use tower_lsp::lsp_types::CompletionItemKind;
 use tree_sitter::Node;
-use workspace::completions_from_workspace;
 
-use crate::lsp::completions::parameter_hints::ParameterHints;
-use crate::lsp::document_context::DocumentContext;
-use crate::lsp::state::WorldState;
+use crate::lsp::completions::builder::CompletionBuilder;
+use crate::lsp::completions::sources::CompletionSource;
 use crate::treesitter::NodeType;
 use crate::treesitter::NodeTypeExt;
 
-pub fn completions_from_composite_sources(
-    context: &DocumentContext,
-    state: &WorldState,
-    parameter_hints: ParameterHints,
-) -> Result<Vec<CompletionItem>> {
-    log::info!("completions_from_composite_sources()");
+/// Aggregator for composite completion sources
+/// Does deduplication and sorting also
+pub struct CompositeCompletionsSource;
 
-    let mut completions: Vec<CompletionItem> = vec![];
-
-    let root = find_pipe_root(context)?;
-
-    // Try argument completions
-    if let Some(mut additional_completions) = completions_from_call(context, root.clone())? {
-        completions.append(&mut additional_completions);
-    }
-
-    // Try pipe completions
-    if let Some(mut additional_completions) = completions_from_pipe(root.clone())? {
-        completions.append(&mut additional_completions);
-    }
-
-    // Try subset completions (`[` or `[[`)
-    if let Some(mut additional_completions) = completions_from_subset(context)? {
-        completions.append(&mut additional_completions);
-    }
-
-    // Call, pipe, and subset completions should show up no matter what when
-    // the user requests completions (this allows them to Tab their way through
-    // completions effectively without typing anything). For the rest of the
-    // general completions, we require an identifier to begin showing
-    // anything.
-    if is_identifier_like(context.node) {
-        completions.append(&mut completions_from_keywords());
-        completions.append(&mut completions_from_snippets());
-        completions.append(&mut completions_from_search_path(context, parameter_hints)?);
-
-        if let Some(mut additional_completions) = completions_from_document(context)? {
-            completions.append(&mut additional_completions);
-        }
-
-        if let Some(mut additional_completions) =
-            completions_from_workspace(context, state, parameter_hints)?
-        {
-            completions.append(&mut additional_completions);
-        }
+impl CompletionSource for CompositeCompletionsSource {
```

Comment:
I think it's a bit unexpected for the composite and unique source routines to implement the `CompletionSource` trait. For instance the `name()` method is never used for those if I'm not missing anything, which I think reveals an asymmetry that makes the aggregators a weird fit for the source trait.

Reply:
Done.